### PR TITLE
fix: redeploy message shown incorrectly when exporting api after creation

### PR DESCRIPTION
**Pull request by julian-robotham-gravitee**
## Issue
First API Export Causes API Desynchronization

https://gravitee.atlassian.net/browse/APIM-1701

## Description

Added functionality to ensure when the API is exported (First export) after being created then the banner requesting the end user to redeploy the API is not shown as no changes have been made.

The Issue was that the string comparison made did not account for the Proxy -> Group -> Endpoint string being compared in an incorrect order see **_Figure 1.1_** - Even though the object is in fact equal the string comparison declared it as out of sync due to the order in which the Object attributes where being compared.

incorrect order compare
**_Figure 1.1_**

The other issue was the null comparison check between the Properties `io/gravitee/definition/model/Properties.java`& Services `io.gravitee.definition.model.services.Services` - although not exactly the same both the deployed API and the API to be deployed have empty values after export and so we have ignored this difference. see **_Figure 1.2_**

null compare
**_Figure 1.2_**

 

## Additional context

Assumptions made:

- null comparison with an empty array / object  should equate to true for Properties `io/gravitee/definition/model/Properties.java`& Services `io.gravitee.definition.model.services.Services`

- the Endpoint comparison only compares the endpoint name in the equals method `io/gravitee/definition/model/Endpoint.java`

- Create and Deploy a Keyless API
- Export API (select all or no options)
- The banner is then shown to the user to redeploy API see **_Figure 2.1_**

API banner on first export
**_Figure 2.1_**

- The correct result is shown in **_Video 2.1_**

https://github.com/gravitee-io/gravitee-api-management/assets/137767084/c0c21d68-c608-4696-adc0-cef4d799e9b9
_**Video 2.1**_

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-mvnrvuuywa.chromatic.com)

🏗️ Your changes can be tested here and will be available soon:
      Console: [https://4726.team-apim.gravitee.dev/console](https://4726.team-apim.gravitee.dev/console)
      Portal: [https://4726.team-apim.gravitee.dev/portal](https://4726.team-apim.gravitee.dev/portal)
      Management-api: [https://4726.team-apim.gravitee.dev/api/management](https://4726.team-apim.gravitee.dev/api/management)
      Gateway v4: [https://4726.team-apim.gravitee.dev](https://4726.team-apim.gravitee.dev)
      Gateway v3: [https://4726.gateway-v3.team-apim.gravitee.dev](https://4726.gateway-v3.team-apim.gravitee.dev)

### CODE REVIEW

**Comment by gaetanmaisse** [comment]
You need to reformat the code with Prettier: `mvn prettier:write -T 2C`

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/processor/SynchronizationService.java [hunk 1]
```diff
@@ -48,8 +53,49 @@ public <T> boolean checkSynchronization(final Class<T> entityClass, final T depl
...
+                                .stream()
+                                .findFirst()
+                                .get()
+                                .getEndpoints()
+                                .stream()
+                                .findFirst()
+                                .get();
+
+                        if (!(deployedEntityProxyGroupEndpoint.equals(entityToDeployProxyGroupEndpoint))) {
+                            return false;
```
Comment (a-cordier) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/processor/SynchronizationService.java]: This causes an issue I think. If I have two endpoints and edit the second one, the banner should show up. With this, it does not.

**Review by a-cordier — DISMISSED** [review]

**Comment by a-cordier** [comment]
Appart from diff comment, couldn't we achieve proper comparison by setting up jackson serialization properly ?

**Comment by a-cordier** [comment]
also, I think there's an issue with the history of the branch

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/processor/SynchronizationService.java (hunk 1 above)
Comment (julian-robotham-gravitee) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/processor/SynchronizationService.java]: New changes show the banner on endpoint changes . Thanks for the input

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/processor/SynchronizationService.java [hunk 2]
```diff
@@ -48,6 +50,23 @@ public <T> boolean checkSynchronization(final Class<T> entityClass, final T depl
                 boolean previousAccessibleState = f.isAccessible();
                 f.setAccessible(true);
                 try {
+                    // FIXME dirty hack to ignore null pointer and empty object comparisions on Properties & Services
```
Comment (a-cordier) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/processor/SynchronizationService.java]: just a thought. How about ensuring properties default to an empty list if null ?

I think that would keep this part of the code cleaner and avoid the FIXME

e.g.

```java
public class Properties implements Serializable {

    @JsonProperty("properties")
    private List<Property> properties = List.of();

    @JsonIgnore
    private Map<String, String> entries = Map.of();

    public void setProperties(List<Property> properties) {
        this.properties = properties == null ? List.of() : properties;

        this.entries = this.properties.stream()
                .collect(
                    Collectors.toMap(
                        Property::getKey,
                        Property::getValue,
                        (v1, v2) -> {
                            throw new RuntimeException(String.format("Duplicate key for values %s and %s", v1, v2));
                        },
                        TemplatedValueHashMap::new
                    )
                );
    }

    public List<Property> getProperties() {
        return properties;
    }

    @JsonIgnore
    public Map<String, String> getValues() {
        return this.entries;
    }
}
```

Comment (a-cordier) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/processor/SynchronizationService.java]: Also, I ran a test with empty service and cannot replicate the issue

Comment (a-cordier) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/processor/SynchronizationService.java]: And maybe for this specific situation (null vs empty properties) we could add a unit test ?

**Comment by a-cordier** [comment]
LGTM but still needs squashing all the commits.

**Comment by julian-robotham-gravitee** [comment]
- Added missing test file to legacy code
- Decreased code complexity  and improved code coverage
- Re-tested and passes local validation

**Comment by julian-robotham-gravitee** [comment]
> LGTM but still needs squashing all the commits.
Done

**Review by jgiovaresco — APPROVED** [review]
